talisker/metrics.py

```diff
@@ -149,45 +153,36 @@ def histogram_sorter(sample):
 
 
 def prometheus_cleanup_worker(pid):
-    """Clean up after a multiprocess worker has died."""
-    if not prometheus_client:
+    """Aggregate dead worker's metrics into a single archive file."""
+    mark_process_dead(pid)
+    prom_dir = os.environ['prometheus_multiproc_dir']
+    worker_files = [
+        'histogram_{}.db'.format(pid),
+        'counter_{}.db'.format(pid),
+    ]
+    paths = [os.path.join(prom_dir, f) for f in worker_files]
+    paths = [p for p in paths if os.path.exists(p)]
+
+    # check at least one worker file exists
+    if not paths:
         return
-    from prometheus_client.multiprocess import mark_process_dead
-
-    try:
-        mark_process_dead(pid)  # just deletes gauge files
-        prom_dir = os.environ['prometheus_multiproc_dir']
-        worker_files = [
-            'histogram_{}.db'.format(pid),
-            'counter_{}.db'.format(pid),
-        ]
-        paths = [os.path.join(prom_dir, f) for f in worker_files]
-        paths = [p for p in paths if os.path.exists(p)]
-
-        # check at least one worker file exists
-        if not paths:
-            return
 
-        histogram_path = os.path.join(prom_dir, histogram_archive)
-        counter_path = os.path.join(prom_dir, counter_archive)
-        metrics = collect(paths + [histogram_path, counter_path])
+    histogram_path = os.path.join(prom_dir, histogram_archive)
+    counter_path = os.path.join(prom_dir, counter_archive)
 
-        tmp_histogram = tempfile.NamedTemporaryFile(delete=False)
-        tmp_counter = tempfile.NamedTemporaryFile(delete=False)
-        write_metrics(metrics, tmp_histogram.name, tmp_counter.name)
+    metrics = collect(paths + [histogram_path, counter_path])
 
-        os.rename(tmp_histogram.name, histogram_path)
-        os.rename(tmp_counter.name, counter_path)
+    tmp_histogram = tempfile.NamedTemporaryFile(delete=False)
+    tmp_counter = tempfile.NamedTemporaryFile(delete=False)
+    write_metrics(metrics, tmp_histogram.name, tmp_counter.name)
 
-        for path in paths:
-            os.unlink(path)
+    # TODO: race condition window starts here
```

Comment:
This'd be a new worker reusing the same PID, right?

Reply:
No, that's a slightly different issue. 

This is:
master: read counter_<pid>.db, and write updated counter_archive file
worker, handling scrape concurrently: read updated counter_archive file, as well as not yet deleted counter_<pid>.db file, and double count them.
master: delete counter_<pid>.db.

It's a small window, but I have seen it in real life. We need a way to make the rename/delete atomic between multiple processes, which probably means a file-system based lock. But it's a separate issue.